hutch_python/obj_config.py

```diff
@@ -0,0 +1,189 @@
+import logging
+from pathlib import Path
+from typing import Callable, Tuple, Union
+
+import yaml
+from ophyd import Kind
+from pcdsdevices.interface import BaseInterface
+
+from . import utils
+
+logger = logging.getLogger(__name__)
+
+
+def update_objs(
+    obj_ns: utils.HelpfulNamespace,
+    entry: str,
+    attrs: dict,
+    fn: Callable[[BaseInterface, Union[dict, list[str]]], None]
+) -> None:
+    """
+    Helper function for updating object namespaces
+    obj_ns is a HelpfulNamespace, which may include other
+    HelpfulNamespace's.  As a result this must be semi-recursive.
+
+    Parameters
+    ----------
+    obj_ns : HelpfulNamespace
+        Namespace of all objects
+
+    entry : str
+        Entry in obj_conf yaml. Name of device or class of devices to edit.
+
+    attrs : list[str]
+        List of attributes to add to whitelist
+
+    fn : Callable[[BaseInterface, dict], None]
+        A function that applies a change to a singular device
+    """
+    dev = getattr(obj_ns, entry, None)
+    if dev:  # look for explicit device
+        fn(dev, attrs)
+
+        return
+
+    # look for devices with class
+    for dev in obj_ns:
+        if isinstance(dev, utils.HelpfulNamespace):
+            update_objs(dev, entry, attrs, fn)
+        # a poor-man's type check, in an effort to avoid importing
+        # every device type again
+        if entry == type(dev).__name__:
+            fn(dev, attrs)
+
+
+def update_whitelist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Update the tab whitelist for ``dev`` with each field in ``attrs``
+
+    Parameters
+    ----------
+    dev : BaseInterface
+        A single object implementing the tab completion features in
+        pcdsdevices.interface.BaseInterface
+
+    attrs : list[str]
+        list of attributes to add to ``dev`` tablist
+    """
+    for att in attrs:
+        dev._tab.add(att)
+
+
+def update_blacklist(dev: BaseInterface, attrs: list[str]) -> None:
+    """
+    Remove each field in ``attrs`` from tab whitelist for ``dev``
```

Comment:
This will fail in the niche case where something is in the tab whitelist by virtue of a regex and the user tries to blacklist it... not sure what to do about it...

Reply:
Are users adding to the regex manually?  It looks to me that the regex is cleared with `add` and `remove`, then rebuilt when `__dir__` is called.  

Are users supplying their own tab completion patterns, and is that the usage pattern we're targeting?  There's the `tab_whitelist` attribute, but adding to that doesn't actually change the whitelist?

Reply:
This is an issue in the tab completion interface in general, I think.  I can take a stab